app/helpers/application_helper.rb

```diff
@@ -70,77 +70,96 @@ def allowed_html_instructions(show_text = true)
-        external_creatorships = creation.external_creatorships.select { |ec| !ec.claimed? }
-        external_creatorships.each do |ec|
-          archivist_pseud = pseuds.select { |p| ec.archivist.pseuds.include?(p) }.first
-          archivists[archivist_pseud] += [ec.author_name]
-        end
+  def byline_text(creation, only_path, text_only: false)
+    # Update Series#expire_byline_cache and Chapter#expire_byline_cache when changing cache key here
+    creators = Rails.cache.fetch(["byline_data", creation.cache_key]) { byline_data(creation) }
+    byline_text_internal(creators, only_path, text_only)
+  end
+
+  def byline_text_uncached(creation, only_path, text_only: false)
+    creators = byline_data(creation)
+    byline_text_internal(creators, only_path, text_only)
+  end
+
+  def byline_text_internal(creators, only_path, text_only)
+    return creators if creators.is_a?(String)
+
+    safe_join(creators.map do |creator|
+      pseud_byline = if text_only
+                       creator[:byline]
+                     else
+                       url_options = { controller: "pseuds", action: "show", user_id: creator[:user], id: creator[:pseud], only_path: only_path }
+                       link_to(creator[:byline], url_options, rel: "author")
+                     end
+
+      if creator[:archivists].empty?
+        pseud_byline
+      else
+        safe_join(creator[:archivists].map do |ext_author|
+          t("application_helper.archivist_byline_html", external_author: ext_author, pseud_byline: pseud_byline)
+        end, t("support.array.words_connector"))
       end
+    end, t("support.array.words_connector"))
+  end
 
-      pseuds.map { |pseud|
-        pseud_byline = text_only ? pseud.byline : pseud_link(pseud, only_path)
-        if archivists[pseud].empty?
-          pseud_byline
-        else
-          archivists[pseud].map { |ext_author|
-            ts("%{ext_author} [archived by %{name}]", ext_author: ext_author, name: pseud_byline)
-          }.join(', ')
-        end
-      }.join(', ').html_safe
+  def byline_data(creation)
+    return creation.author if creation.respond_to?(:author)
+
+    pseuds = @preview_mode ? creation.pseuds_after_saving : creation.pseuds.to_a # rubocop:disable Rails/HelperInstanceVariable
+    pseuds = pseuds.flatten.uniq.sort
+
+    archivists = Hash.new []
+    if creation.is_a?(Work)
+      external_creatorships = creation.external_creatorships.reject(&:claimed?)
+      external_creatorships.each do |ec|
+        archivist_pseud = pseuds.find { |p| ec.archivist.pseuds.include?(p) }
+        archivists[archivist_pseud] += [ec.author_name]
+      end
     end
-  end
 
-  def pseud_link(pseud, only_path = true)
-    if only_path
-      link_to(pseud.byline, user_pseud_path(pseud.user, pseud), rel: "author")
-    else
-      link_to(pseud.byline, user_pseud_url(pseud.user, pseud), rel: "author")
+    pseuds.map do |pseud|
+      {
+        byline: pseud.byline,
+        pseud: pseud.to_param,
```

Comment:
`to_param` should be called by `link_to`/`url_for`, but do you call it here so they're more cacheable strings? Worth a comment as well.